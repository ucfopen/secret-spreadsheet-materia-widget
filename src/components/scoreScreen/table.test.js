import React from 'react';
import ScoreTable from './table';
import renderer from 'react-test-renderer';
import { shallow } from '../../enzyme';

const props = () => {
	return {
		spreadsheet: false,
		dimensions: {
			rows: 2,
			columns: 2
		},
		header: false,
		qset: [
			[
				{
					"materiaType": `question`,
					"id": null,
					"type": `QA`,
					"questions": [
						{
							"text": `Show 1`
						}
					],
					"answers": [
						{
							"id": null,
							"text": `Show 1`,
							"value": 100
						}
					],
					"options": {
						"blank": false,
						"position": {
							"row": 0,
							"column": 0
						}
					}
				},
				{
					"materiaType": `question`,
					"id": null,
					"type": `QA`,
					"questions": [
						{
							"text": `Hide 1`
						}
					],
					"answers": [
						{
							"id": null,
							"text": `Hide 1`,
							"value": 100
						}
					],
					"options": {
						"blank": true,
						"position": {
							"row": 0,
							"column": 1
						}
					}
				}
			],
			[
				{
					"materiaType": `question`,
					"id": null,
					"type": `QA`,
					"questions": [
						{
							"text": `Hide 2`
						}
					],
					"answers": [
						{
							"id": null,
							"text": `Hide 2`,
							"value": 100
						}
					],
					"options": {
						"blank": true,
						"position": {
							"row": 1,
							"column": 0
						}
					}
				},
				{
					"materiaType": `question`,
					"id": null,
					"type": `QA`,
					"questions": [
						{
							"text": `Show 2`
						}
					],
					"answers": [
						{
							"id": null,
							"text": `Show 2`,
							"value": 100
						}
					],
					"options": {
						"blank": false,
						"position": {
							"row": 1,
							"column": 1
						}
					}
				}
			]
		],
		positions: new Set([`row0 column1`, `row1 column0`]),
		leftAlign: false,
		scoreTable: [
			{
				data: [
					`Hide 1`, `Hide 1`, `Hide 1`, `row0 column1`
				],
				data_style: [
					`question`, `response`, `answer`, `position`
				],
				display_score: true,
				feedback: null,
				graphic: `score`,
				score: 100,
				style: `full-value`,
				symbol: `%`,
				tag: `div`,
				type: `SCORE_QUESTION_ANSWERED`
			},
			{
				data: [
					`Hide 2`, `wrong`, `Hide 2`, `row1 column0`
				],
				data_style: [
					`question`, `response`, `answer`, `position`
				],
				display_score: true,
				feedback: null,
				graphic: `score`,
				score: 100,
				style: `full-value`,
				symbol: `%`,
				tag: `div`,
				type: `SCORE_QUESTION_ANSWERED`
			}
		]
	};
};

const customProps = (changes={}) => {
	const newProps = props();

	Object.defineProperties(newProps, changes);

	return newProps;
};

describe(`ScoreTable`, () => {
	test(`Rendered with header, spreadsheet, centered`, () => {
		const newProps = customProps({
			header: {value: true},
			spreadsheet: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with header, spreadsheet, left aligned`, () => {
		const newProps = customProps({
			header: {value: true},
			spreadsheet: {value: true},
			leftAlign: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with header, no spreadsheet, centered`, () => {
		const newProps = customProps({
			header: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with header, no spreadsheet, left aligned`, () => {
		const newProps = customProps({
			header: {value: true},
			leftAlign: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with no header, spreadsheet, centered`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with no header, spreadsheet, left aligned`, () => {
		const newProps = customProps({
			leftAlign: {value: true},
			spreadsheet: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with no header, no spreadsheet, centered`, () => {
		const newProps = props();

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`Rendered with no header, no spreadsheet, left aligned`, () => {
		const newProps = customProps({
			leftAlign: {value: true}
		});

		const component = renderer.create(<ScoreTable {... newProps} />);
		const tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test(`handleChange works going from not checked to checked`, () => {
		const event = {
			target: {
				checked: true
			}
		};
		const newProps = props();

		const component = shallow(<ScoreTable {... newProps} />);
		component.setState({checked: false});
		component.instance().handleChange(event);

		expect(component.state([`checked`])).toBeTruthy();

		// cleanup
		component.unmount();
	});

	test(`handleChange works going from checked to not checked`, () => {
		const event = {
			target: {
				checked: false
			}
		};
		const newProps = props();

		const component = shallow(<ScoreTable {... newProps} />);
		component.setState({checked: true});
		component.instance().handleChange(event);

		expect(component.state([`checked`])).toBeFalsy();

		// cleanup
		component.unmount();
	});

	test(`convertNumberToLetters number less than 0`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);
		const result = component.instance().convertNumberToLetters(-1);

		expect(typeof result).toBe(`string`);
		expect(result).toBe(`A`);

		// cleanup
		component.unmount();
	});

	test(`convertNumberToLetters number is 0`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);
		const result = component.instance().convertNumberToLetters(0);

		expect(typeof result).toBe(`string`);
		expect(result).toBe(`A`);

		// cleanup
		component.unmount();
	});

	test(`convertNumberToLetters number is greater than 0`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);
		const result = component.instance().convertNumberToLetters(1);

		expect(typeof result).toBe(`string`);
		expect(result).toBe(`B`);

		// cleanup
		component.unmount();
	});

	test(`convertNumberToLetters number is very large`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);
		const result = component.instance().convertNumberToLetters(1000000000);

		expect(typeof result).toBe(`string`);
		expect(result).toBe(`CGEHTYM`);

		// cleanup
		component.unmount();
	});

	test(`convertNumberToLetters gets an average number`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);
		const result = component.instance().convertNumberToLetters(40);

		expect(typeof result).toBe(`string`);
		expect(result).toEqual(`AO`);

		//cleanup
		component.unmount();
	});

	test(`convertNumberToLetters doesn't get a number`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});
		const realError = console.error;
		console.error = jest.fn();

		const component = shallow(<ScoreTable {... newProps} />);
		const result = component.instance().convertNumberToLetters(`test`);

		expect(console.error).toHaveBeenCalledTimes(1);
		expect(result).toBe(NaN);

		// cleanup
		component.unmount();
		console.error = realError;
	});

	test(`generateBody with start at 0 and the answer is not in the scoreTable`, () => {
		const newProps = customProps({
			scoreTable: {value: []}
		});
		const realError = console.error;
		console.error = jest.fn();

		const component = shallow(<ScoreTable {... newProps} />);
		component.instance().generateBody();

		// four times because three times is with the original generation
		expect(console.error).toHaveBeenCalledTimes(4);

		// cleanup
		component.unmount();
		console.error = realError;
	});

	test(`Toggle show answer checkbox with click`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);

		expect(component.instance().state.checked).toBeFalsy();
		component.find(`.show-answer-checkbox`).simulate(`Click`);
		expect(component.instance().state.checked).toBeTruthy();

		//cleanup
		component.unmount();
	});

	test(`Toggle show answer checkbox with Enter`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);

		expect(component.instance().state.checked).toBeFalsy();
		component.find(`.show-answer-checkbox`).simulate(`keyDown`, {key: `Enter`});
		expect(component.instance().state.checked).toBeTruthy();

		//cleanup
		component.unmount();
	});

	test(`Can not toggle show answer checkbox with key other than Enter`, () => {
		const newProps = customProps({
			spreadsheet: {value: true}
		});

		const component = shallow(<ScoreTable {... newProps} />);

		expect(component.instance().state.checked).toBeFalsy();
		component.find(`.show-answer-checkbox`).simulate(`keyDown`, {key: `a`});
		expect(component.instance().state.checked).toBeFalsy();

		//cleanup
		component.unmount();
	});
});
