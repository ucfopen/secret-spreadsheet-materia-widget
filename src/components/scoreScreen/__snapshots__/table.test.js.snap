// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreTable Rendered with header, no spreadsheet, centered 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <thead>
          <tr
            className={null}
            id="row0"
          >
            <th
              className="header"
              id="cell0"
            >
              Show 1
            </th>
            <th
              className="header"
              id="cell1"
            >
              Hide 1
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="row1"
          >
            <td
              className="wrong"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className=""
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with header, no spreadsheet, left aligned 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <thead>
          <tr
            className="leftAlign"
            id="row0"
          >
            <th
              className="header"
              id="cell0"
            >
              Show 1
            </th>
            <th
              className="header"
              id="cell1"
            >
              Hide 1
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="row1"
          >
            <td
              className="wrong leftAlign"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className="leftAlign"
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with header, spreadsheet, centered 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <thead>
          <tr
            id="column-labels"
          >
            <th
              className="label skinny"
              id="col-label-0"
            />
            <th
              className="label"
              id="col-label-1"
            >
              A
            </th>
            <th
              className="label"
              id="col-label-2"
            >
              B
            </th>
          </tr>
          <tr
            className={null}
            id="row0"
          >
            <th
              className="label skinny"
              id="row-label-1"
            >
              1
            </th>
            <th
              className="header"
              id="cell0"
            >
              Show 1
            </th>
            <th
              className="header"
              id="cell1"
            >
              Hide 1
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="row1"
          >
            <td
              className="label skinny"
              id="row-label-2"
            >
              2
            </td>
            <td
              className="wrong"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className=""
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with header, spreadsheet, left aligned 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <thead>
          <tr
            id="column-labels"
          >
            <th
              className="label skinny leftAlign"
              id="col-label-0"
            />
            <th
              className="label"
              id="col-label-1"
            >
              A
            </th>
            <th
              className="label"
              id="col-label-2"
            >
              B
            </th>
          </tr>
          <tr
            className="leftAlign"
            id="row0"
          >
            <th
              className="label skinny leftAlign"
              id="row-label-1"
            >
              1
            </th>
            <th
              className="header"
              id="cell0"
            >
              Show 1
            </th>
            <th
              className="header"
              id="cell1"
            >
              Hide 1
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="row1"
          >
            <td
              className="label skinny leftAlign"
              id="row-label-2"
            >
              2
            </td>
            <td
              className="wrong leftAlign"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className="leftAlign"
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with no header, no spreadsheet, centered 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <tbody>
          <tr
            id="row0"
          >
            <td
              className=""
              id="cell0"
            >
              Show 1
            </td>
            <td
              className="right different"
              id="cell1"
            >
              <p>
                Hide 1
              </p>
            </td>
          </tr>
          <tr
            id="row1"
          >
            <td
              className="wrong"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className=""
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with no header, no spreadsheet, left aligned 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <tbody>
          <tr
            id="row0"
          >
            <td
              className="leftAlign"
              id="cell0"
            >
              Show 1
            </td>
            <td
              className="right leftAlign different"
              id="cell1"
            >
              <p>
                Hide 1
              </p>
            </td>
          </tr>
          <tr
            id="row1"
          >
            <td
              className="wrong leftAlign"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className="leftAlign"
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with no header, spreadsheet, centered 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <thead>
          <tr
            id="column-labels"
          >
            <th
              className="label skinny"
              id="col-label-0"
            />
            <th
              className="label"
              id="col-label-1"
            >
              A
            </th>
            <th
              className="label"
              id="col-label-2"
            >
              B
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="row0"
          >
            <td
              className="label skinny"
              id="row-label-1"
            >
              1
            </td>
            <td
              className=""
              id="cell0"
            >
              Show 1
            </td>
            <td
              className="right different"
              id="cell1"
            >
              <p>
                Hide 1
              </p>
            </td>
          </tr>
          <tr
            id="row1"
          >
            <td
              className="label skinny"
              id="row-label-2"
            >
              2
            </td>
            <td
              className="wrong"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className=""
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;

exports[`ScoreTable Rendered with no header, spreadsheet, left aligned 1`] = `
<main>
  <h3>
    Answers
  </h3>
  <label
    id="check-label"
  >
    Show correct answers
    <div
      aria-checked="false"
      aria-labelledby="check-label"
      className="show-answer-checkbox"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="checkbox"
      tabIndex={0}
    >
      <svg
        height="22px"
        viewBox="0 0 28 28"
        width="22px"
      >
        <path
          d="M0 0v28h28V0H0zm24 24H4V4h20v20z"
        />
      </svg>
    </div>
  </label>
  <div
    className="table-surround"
  >
    <div>
      <table>
        <thead>
          <tr
            id="column-labels"
          >
            <th
              className="label skinny leftAlign"
              id="col-label-0"
            />
            <th
              className="label"
              id="col-label-1"
            >
              A
            </th>
            <th
              className="label"
              id="col-label-2"
            >
              B
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="row0"
          >
            <td
              className="label skinny leftAlign"
              id="row-label-1"
            >
              1
            </td>
            <td
              className="leftAlign"
              id="cell0"
            >
              Show 1
            </td>
            <td
              className="right leftAlign different"
              id="cell1"
            >
              <p>
                Hide 1
              </p>
            </td>
          </tr>
          <tr
            id="row1"
          >
            <td
              className="label skinny leftAlign"
              id="row-label-2"
            >
              2
            </td>
            <td
              className="wrong leftAlign"
              id="cell2"
            >
              <p>
                wrong
              </p>
            </td>
            <td
              className="leftAlign"
              id="cell3"
            >
              Show 2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</main>
`;
